### PR TITLE
reduce IPC platform driver API

### PR DESCRIPTION
**Pull request by lyakh**
Most importantly this patch series removes 2 IPC platform driver methods: .get_ready and .cmd_done, additionally it contains some simple improvements. Removing those methods is *not* a completely "lossless" process, e.g. some register access operations had to be swapped. That *shouldn't* cause problems, but it has to be thoroughly tested for any regressions. I've only tested on Up^2 with a HifiBerry DAC+ on its RCA output.

### CODE REVIEW

File: sound/soc/sof/intel/bdw.c [hunk 1]
```diff
@@ -489,13 +489,12 @@ static int bdw_get_reply(struct snd_sof_dev *sdev, struct snd_sof_ipc_msg *msg)
 	} else {
```
Comment (plbossart) [inline sound/soc/sof/intel/bdw.c:499]: @lyakh please add a more complete prefix in your patches, e.g. for this one this should be ASoC: SOF: Intel since you are only touching intel-specific patches.

File: sound/soc/sof/hw-spi.c [hunk 2]
```diff
@@ -167,20 +167,6 @@ static irqreturn_t spi_irq_handler(int irq __maybe_unused, void *context)
...
-	/*
-	 * if reply. Handle Immediate reply from DSP Core and set DSP
-	 * state to ready
-	 */
-
-	/* if cmd, Handle messages from DSP Core */
-
-	return IRQ_HANDLED;
-}
-
```
Comment (plbossart) [inline sound/soc/sof/hw-spi.c:183]: @lyakh can you do me a favor and not bundle things aren't ready for upstream with things that are. Specifically Haswell/SKL/KBL/Compressed/SPI are not shared upstream just yet, so it helps if your patches follow this split.

File: sound/soc/sof/intel/byt.c [hunk 3]
```diff
@@ -387,13 +387,12 @@ static int byt_get_reply(struct snd_sof_dev *sdev, struct snd_sof_ipc_msg *msg)
 	} else {
 		/* reply correct size ? */
 		if (reply.hdr.size != msg->reply_size) {
-			dev_err(sdev->dev, "error: reply expected 0x%zx got 0x%x bytes\n",
+			dev_err(sdev->dev, "error: reply expected 0x%zx got %u bytes\n",
```
Comment (lgirdwood) [inline sound/soc/sof/intel/byt.c]: Does %u make it obvious to the user that it's base16 ?

Comment (plbossart) [reply sound/soc/sof/intel/byt.c]: %u is for size_t and will print in decimal. Andy commented that the sizes should be decimal, not hex and I agree.

Comment (lyakh) [reply sound/soc/sof/intel/byt.c]: Oh, I missed that there are two sizes printed in this one, so both have to be made %u. And for size_t it might be %zu, I'll have to double check.

File: sound/soc/sof/hw-spi.c (hunk 2 above)
Comment (lyakh) [reply sound/soc/sof/hw-spi.c:183]: sorry, what branch should I be making patches against? I was using topic/sof-dev, should I use integration/... instead?

Comment (ranj063) [reply sound/soc/sof/hw-spi.c:183]: @lyakh we should use topic/sof-dev but if you look at the patches posted upstream HSW/KBL/SKL/Compressed/SPI are not included in them. So it will be easier to maintain the squash/rebase cycles if you split for ex. your commit: "ASoC: SOF: eliminate redundant mailbox reads" into 2. One intended for the files in the upstream patches and another one for those not in it (hsw.c etc).

Comment (lyakh) [reply sound/soc/sof/hw-spi.c:183]: I see, but how should I know which ones are and which ones aren't included? I think we need a separate branch, really.

Comment (plbossart) [reply sound/soc/sof/hw-spi.c:183]: @lyakh check topic/sof-dev-rebase, everything after 'add build support' is not upstream yet. Ideally your patches should apply cleanly on the initial batch. Thanks!

Comment (lyakh) [reply sound/soc/sof/hw-spi.c:183]: @plbossart let me take a semi-qualified guess: is this the commit, that I should be branching from?
> commit af23c43d6a9148df393735a9fe3c8a45e8a6435d
> Author: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>
> Date:   Mon May 21 18:19:44 2018 -0500
> 
>     ASoC: Intel: boards: initial support for Hifiberry Digiplus w/ WM8804

**Comment by plbossart** [comment]
> @plbossart <https://github.com/plbossart> let me take a semi-qualified 
> guess: is this the commit, that I should be branching from?

the last patch submitted upstream is c12f34018c9a ('ASoC: Intel: select 
relevant machine drivers for SOF')

**Comment by lyakh** [comment]
@plbossart I forked from an earlier commit that didn't contain hsw, spi, etc., rebased my patches on top of it and pushed it. Now gh is trying to merge this to the head, which creates a mess, of course. Can you maybe just pull / cherry pick from that my branch?

**Comment by plbossart** [comment]
> @plbossart I forked from an earlier commit that didn't contain hsw, spi, etc., rebased my patches on top of it and pushed it. Now gh is trying to merge this to the head, which creates a mess, of course. Can you maybe just pull / cherry pick from that my branch?

sorry no, I can only review and use the web interface this week. What's the issue with splitting the patches in a part that will apply on the v5 set and one that applies on top of topic/sof-dev?

**Comment by lyakh** [comment]
> > @plbossart I forked from an earlier commit that didn't contain hsw, spi, etc., rebased my patches on top of it and pushed it. Now gh is trying to merge this to the head, which creates a mess, of course. Can you maybe just pull / cherry pick from that my branch?
> 
> sorry no, I can only review and use the web interface this week. What's the issue with splitting the patches in a part that will apply on the v5 set and one that applies on top of topic/sof-dev?

If there were a branch somewhere with exactly v5 there would be no issue at all. But if I understood correctly I was supposed to make patches that apply "to a branch but not to its head but N patches before the head." That's what I've done here, but when pushing and making a PR the github produced a PR with 47 commits and a mountain of conflicts. Maybe I've done something wrong, but it *really* would have been much easier to have a clean branch with just v5 and nothing else in it.

**Comment by plbossart** [comment]
topic/sof-def-rebase branch, commit e92423f7a73cee5c1f34dd016c2f7c96a659f39b contains exactly the v5 patches.

**Comment by lyakh** [comment]
> topic/sof-def-rebase branch, commit [e92423f](https://github.com/thesofproject/linux/commit/e92423f7a73cee5c1f34dd016c2f7c96a659f39b) contains exactly the v5 patches.

right, and it's 24 commits behind the topic/sof-dev-rebase head, which is exactly the problem

**Comment by lyakh** [comment]
@plbossart how do we progress with this?

**Comment by plbossart** [comment]
> @plbossart how do we progress with this?

I really don't get the issue. Split the patches in two, and make sure the first patches apply on top of v5, and the second batch on top of sof-dev, then submit the whole thing as one PR.

**Comment by lyakh** [comment]
> > @plbossart how do we progress with this?
> 
> I really don't get the issue. Split the patches in two, and make sure the first patches apply on top of v5, and the second batch on top of sof-dev, then submit the whole thing as one PR.

@plbossart ah, so you want me to split my patches into two lots so that the first one *can* apply to v5 and the second one applies on top of sof-dev+my first lot, and then (hopefully the first lot will also apply on top of sof-dev) apply both lots in that order on top of sof-dev and push to you. After that you'll again split them...

**Comment by plbossart** [comment]
> > > @plbossart how do we progress with this?
> > 
> > 
> > I really don't get the issue. Split the patches in two, and make sure the first patches apply on top of v5, and the second batch on top of sof-dev, then submit the whole thing as one PR.
> 
> @plbossart ah, so you want me to split my patches into two lots so that the first one _can_ apply to v5 and the second one applies on top of sof-dev+my first lot, and then (hopefully the first lot will also apply on top of sof-dev) apply both lots in that order on top of sof-dev and push to you. After that you'll again split them...

I would actually cherry-pick the first lot to push it as either incremental patches to alsa-devel if Mark merges the v5, or squash them into a v6 otherwise. You will save me time with this split, thanks in advance!

**Comment by lyakh** [comment]
@plbossart I think something went wrong with github links, both your references to https://github.com/thesofproject/linux/commit/c12f34018c9af1baa08d389ced175dad6328b1ee and https://github.com/thesofproject/linux/commit/e92423f7a73cee5c1f34dd016c2f7c96a659f39b point to a commit from 2017.

**Comment by plbossart** [comment]
> @plbossart I think something went wrong with github links, both your references to [c12f340](https://github.com/thesofproject/linux/commit/c12f34018c9af1baa08d389ced175dad6328b1ee) and [e92423f](https://github.com/thesofproject/linux/commit/e92423f7a73cee5c1f34dd016c2f7c96a659f39b) point to a commit from 2017.

the commit are quite old, i've been squashing for 1.5 years now. It's the branch that matters.

**Comment by lyakh** [comment]
@plbossart I think this should do it now: the first 7 commits are for upstream, the remaining 5 are for sof-dev

File: sound/soc/sof/intel/bdw.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/intel/bdw.c:499]: @lyakh this comment still applies. ASoC: SOF: Intel prefix please, it makes things easier to track for reviewers (some of whom may not care about Intel specifics)

File: sound/soc/sof/intel/byt.c [hunk 4]
```diff
@@ -396,7 +396,7 @@ static int byt_get_reply(struct snd_sof_dev *sdev, struct snd_sof_ipc_msg *msg)
 	}
 
 	/* read the message */
-	if (msg->msg_data && size > 0)
+	if (size > 0)
```
Comment (plbossart) [inline sound/soc/sof/intel/byt.c]: comment on subject: ASoC: SOF: Intel again

File: sound/soc/sof/sof-priv.h [hunk 5]
```diff
@@ -359,6 +358,7 @@ struct snd_sof_dev {
 	struct snd_sof_mailbox dsp_box;		/* DSP initiated IPC */
 	struct snd_sof_mailbox host_box;	/* Host initiated IPC */
 	struct snd_sof_mailbox stream_box;	/* Stream position update */
+	struct snd_sof_ipc_msg *msg;
```
Comment (plbossart) [inline sound/soc/sof/sof-priv.h:360]: ASoC: SOF: ipc: for the subject line please

File: sound/soc/sof/intel/apl.c [hunk 6]
```diff
@@ -51,7 +51,6 @@ const struct snd_sof_dsp_ops sof_apl_ops = {
 	/* ipc */
 	.send_msg	= hda_dsp_ipc_send_msg,
 	.fw_ready	= hda_dsp_ipc_fw_ready,
-	.cmd_done	= hda_dsp_ipc_cmd_done,
```
Comment (plbossart) [inline sound/soc/sof/intel/apl.c:55]: ASoC: SOF: ipc prefix

File: sound/soc/sof/core.c [hunk 7]
```diff
@@ -231,7 +231,7 @@ int snd_sof_create_page_table(struct snd_sof_dev *sdev,
 		 *    x = (pfn[i+1] << 4) | (pfn[i] & 0xf)
 		 * 4. put x at offset (current location + 2) in LE byte order
 		 * 5. increment current location by 5 bytes, increment i by 2
-		 * 6. continue to (1)
+		 * 6. continue to (2)
```
Comment (plbossart) [inline sound/soc/sof/core.c]: ASoC: SOF: core: prefix for subject, this isn't really related to the IPC.

File: sound/soc/sof/intel/hsw.c [hunk 8]
```diff
@@ -490,13 +490,12 @@ static int hsw_get_reply(struct snd_sof_dev *sdev, struct snd_sof_ipc_msg *msg)
...
-			dev_err(sdev->dev, "error: reply expected 0x%zx got 0x%x bytes\n",
+			dev_err(sdev->dev, "error: reply expected 0x%zx got %u bytes\n",
 				msg->reply_size, reply.hdr.size);
-			size = msg->reply_size;
 			ret = -EINVAL;
-		} else {
-			size = reply.hdr.size;
 		}
+
+		size = msg->reply_size;
```
Comment (plbossart) [inline sound/soc/sof/intel/hsw.c]: ASoC: SOF: hsw: simplify .get_reply() implementations

File: sound/soc/sof/hw-spi.c [hunk 9]
```diff
@@ -167,20 +167,6 @@ static irqreturn_t spi_irq_handler(int irq __maybe_unused, void *context)
 	return IRQ_NONE;
 }
 
-static irqreturn_t spi_irq_thread(int irq __maybe_unused, void *context __maybe_unused)
```
Comment (plbossart) [inline sound/soc/sof/hw-spi.c:170]: I am probably getting on your nerves but even this needs to be split in 3. My topic/sof-def-rebase have haswell, SKL and SPI added in 3 separate steps.

File: sound/soc/sof/intel/bdw.c [hunk 10]
```diff
@@ -489,13 +489,12 @@ static int bdw_get_reply(struct snd_sof_dev *sdev, struct snd_sof_ipc_msg *msg)
...
 		/* reply correct size ? */
 		if (reply.hdr.size != msg->reply_size) {
```
Comment (dbaluta) [inline sound/soc/sof/intel/bdw.c:501]: So, besides removing the else clause your patch also modifies the format for dev_err.  I think that should be the subject of another patch.

**Comment by lyakh** [comment]
@plbossart I've got the next lot of IPC patches, basically just removing .mailbox_read() and .mailbox_write(), should I wait until this is merged or push on top?

**Comment by wenqingfu** [comment]
regression reported on #773 narrowed down to this PR.
